### PR TITLE
Pauli error support for Clifford device

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Adds the support to use noise channels with the Clifford device 

**Description of the Change:** This PR adds the support for using `qml.PauliError`, `qml.BitFlip`, `qml.PhaseFlip` and `qml.DepolarizingChannel`. Based on these channels, we compute noisy samples that would then be used for computing the stochastic measurement results. 

**Benefits:** Users can add Pauli noise to their Clifford circuits. 

**Possible Drawbacks:** Error channels are supported only with finite shots. 

**Related GitHub Issues:**

### CODE REVIEW

**Comment by obliviateandsurrender** [comment]
[sc-43359]

File: pennylane/devices/default_clifford.py [hunk 1]
```diff
@@ -81,7 +81,8 @@
 }
 
 # Clifford gates
-_GATE_OPERATIONS = {
+# Error channels
```
Comment (timmysilv) [inline pennylane/devices/default_clifford.py:84]:
```suggestion
# Clifford gates and error channels
```

File: pennylane/devices/default_clifford.py [hunk 2]
```diff
@@ -540,6 +553,8 @@ def simulate(
         global_phase_ops = []
         for op in circuit.operations[use_prep_ops:]:
             gate, wires = _pl_to_stim(op)
+            if isinstance(op, qml.operation.Channel) and not circuit.shots:
+                raise ValueError(f"{self.name} supports error channels only with finite shots.")
```
Comment (timmysilv) [inline pennylane/devices/default_clifford.py:557]: we'd like to hit this at preprocess-time, not execution-time. Can you register this as a pre-processing transform? wondering if we can sneak this into `decompose` somehow... might be possible with another PR happening in tandem that has a custom stopping condition for finite-shot mode. otherwise just a new transform is fine

**Review by timmysilv — COMMENTED** [review]
looks great!

File: pennylane/devices/default_clifford.py [hunk 3]
```diff
@@ -102,28 +103,53 @@
...
+    return obs.name in _OBSERVABLES_MAP
+
+
+@qml.transform
+def _validate_channels(tape, stopping_condition, name="device"):
+    """Validates the channels for a circuit."""
+    for op in tape.operations:
+        if isinstance(op, qml.operation.Channel):
+            if not stopping_condition(op):
+                raise qml.DeviceError(f"Channel {repr(op)} not supported on {name}")
```
Comment (timmysilv) [inline pennylane/devices/default_clifford.py:129]: I feel a little funny about this. `decompose` runs before with the same stopping condition, so we should theoretically never hit this line (these should either be decomposed or it should have failed). It can only happen if `check_clifford=False`, which is already being sneaky and letting the user override some things.

After typing this, I think my proposal is either:
1. remove `check_clifford` (I think we talked enough about this, and we're keeping it, soooo option 2 might be best)
2. only validate channels if `check_clifford` is True, and remove the `stopping_condition` from this transform. If the user wants to skip it, let them skip it!

File: pennylane/devices/default_clifford.py [hunk 4]
```diff
@@ -102,28 +103,51 @@
...
+    return obs.name in _OBSERVABLES_MAP
+
+
+@qml.transform
+def _validate_channels(tape, name="device"):
+    """Validates the channels for a circuit."""
+    for op in tape.operations:
+        if isinstance(op, qml.operation.Channel):
+            if not tape.shots:
+                raise qml.DeviceError(f"Channel not supported on {name} without finite shots.")
```
Comment (timmysilv) [inline pennylane/devices/default_clifford.py:129]: style nitpick:
```suggestion
    if not tape.shots and any(isinstance(op, qml.operation.Channel) for op in tape.operations):
        raise qml.DeviceError(f"Channel not supported on {name} without finite shots.")
```

**Review by timmysilv — APPROVED** [review]
🎉 

should we merge this first, or wait til the target is in master?

**Comment by obliviateandsurrender** [comment]
I would wait until the target becomes `master`.

**Comment by trbromley** [comment]
No blockers from me, great to have this in!

File: pennylane/devices/default_clifford.py [hunk 5]
```diff
@@ -103,28 +104,49 @@
     "StatePrep": None,
     "Snapshot": None,
     "Barrier": None,
+    "PauliError": "CORRELATED_ERROR",
+    "BitFlip": "X_ERROR",
+    "PhaseFlip": "Z_ERROR",
+    "DepolarizingChannel": "DEPOLARIZE1",
```
Comment (soranjh) [inline pennylane/devices/default_clifford.py:110]: Are these the only error channels stim supports?

File: pennylane/devices/default_clifford.py [hunk 6]
```diff
@@ -103,28 +104,49 @@
...
     """Specifies whether an observable is accepted by ``DefaultClifford``."""
-    return obs.name in _MEAS_OBSERVABLES
+    return obs.name in _OBSERVABLES_MAP
+
+
+@qml.transform
+def _validate_channels(tape, name="device"):
+    """Validates the channels for a circuit."""
+    if not tape.shots and any(isinstance(op, qml.operation.Channel) for op in tape.operations):
+        raise qml.DeviceError(f"Channel not supported on {name} without finite shots.")
```
Comment (soranjh) [inline pennylane/devices/default_clifford.py:128]: [non-blocking] This could be two separate errors.

File: pennylane/devices/default_clifford.py [hunk 7]
```diff
@@ -306,6 +328,35 @@ def circuit(state):
         >>> circuit(basis_states[2])
         tensor(0.0, requires_grad=True)
 
+    .. details::
+        :title: Error Channels
+
+        This device supports the ``finite-shot`` execution of the quantum circuits with
```
Comment (soranjh) [inline pennylane/devices/default_clifford.py:334]:
```suggestion
        This device supports the ``finite-shot`` execution of quantum circuits with
```

File: pennylane/devices/default_clifford.py [hunk 8]
```diff
@@ -306,6 +328,35 @@ def circuit(state):
...
+        the following error channels that add ``Pauli noise``, allowing for one to perform
+        any sampling-based `measurements <https://github.com/quantumlib/stim/>`_.
```
Comment (soranjh) [inline pennylane/devices/default_clifford.py:336]: [optional] Can remove the last part:
"allowing for one to perform any sampling-based `measurements <https://github.com/quantumlib/stim/>`_."

File: tests/devices/test_default_clifford.py [hunk 9]
```diff
@@ -602,6 +620,67 @@ def circuit():
...
+        qml.BitFlip(0.1, wires=[0]),
+        qml.BitFlip(0.5, wires=[2]),
+        qml.PhaseFlip(0.3, wires=[1]),
+        qml.PhaseFlip(0.5, wires=[3]),
+        qml.DepolarizingChannel(0.2, wires=[0]),
+        qml.DepolarizingChannel(0.75, wires=[2]),
+    ],
+)
+def test_meas_noisy_distribution(channel_op):
+    """Test error is raised when noisy circuit are executed on Clifford device analytically."""
```
Comment (soranjh) [inline tests/devices/test_default_clifford.py:665]: Needs update.

**Review by soranjh — CHANGES_REQUESTED** [review]
Looks good, left some minor comments.

File: pennylane/devices/default_clifford.py (hunk 5 above)
Comment (obliviateandsurrender) [reply pennylane/devices/default_clifford.py:110]: `Stim` support many more, but we don't have a corresponding (straightforward) `Channel` mapping for them in PennyLane.

File: pennylane/devices/default_clifford.py (hunk 6 above)
Comment (obliviateandsurrender) [reply pennylane/devices/default_clifford.py:128]: These two conditions are kept combined because we cannot support `Channels` only if the first condition is met i.e. analytical case. One can use channels if they are in the `finite-shot` region. On the other hand, if an unsupported channel is given, it is already being checked before the following check is performed in the `preprocess` method.

**Review by soranjh — APPROVED** [review]

File: pennylane/devices/default_clifford.py [hunk 10]
```diff
@@ -307,6 +336,36 @@ def circuit(state):
         >>> circuit(basis_states[2])
         tensor(0.0, requires_grad=True)
 
+    .. details::
+        :title: Error Channels
+        :href: clifford-errors
+
+        This device supports the ``finite-shot`` execution of quantum circuits with
+        the following error channels that add ``Pauli noise``, allowing for one to perform
```
Comment (trbromley) [inline pennylane/devices/default_clifford.py:344]:
```suggestion
        the following error channels that add Pauli noise, allowing for one to perform
```

File: pennylane/devices/default_clifford.py [hunk 11]
```diff
@@ -307,6 +336,36 @@ def circuit(state):
         >>> circuit(basis_states[2])
         tensor(0.0, requires_grad=True)
 
+    .. details::
+        :title: Error Channels
+        :href: clifford-errors
+
+        This device supports the ``finite-shot`` execution of quantum circuits with
```
Comment (trbromley) [inline pennylane/devices/default_clifford.py:343]:
```suggestion
        This device supports the finite-shot execution of quantum circuits with
```

File: pennylane/devices/default_clifford.py [hunk 12]
```diff
@@ -307,6 +336,36 @@ def circuit(state):
...
+        the following error channels that add ``Pauli noise``, allowing for one to perform
+        any sampling-based `measurements <https://github.com/quantumlib/stim/>`_.
```
Comment (trbromley) [inline pennylane/devices/default_clifford.py:345]: Not sure if this is the right URL. Maybe let's just remove the hyperlink
```suggestion
        any sampling-based measurements.
```

File: pennylane/devices/default_clifford.py [hunk 13]
```diff
@@ -307,6 +336,36 @@ def circuit(state):
...
+            def circuit():
+                qml.Hadamard(wires=[0])
+                for idx in range(num_wires):
+                    qml.CNOT(wires=[idx, idx+1])
+                qml.BitFlip(0.2, wires=[1])
+                return qml.counts()
+
+        >>> circuit()
+        {'0000': 417, '0100': 95, '1011': 104, '1111': 408}
+
```
Comment (trbromley) [inline pennylane/devices/default_clifford.py:368]: Nice!!

**Review by trbromley — APPROVED** [review]
Thanks @obliviateandsurrender, approved once my suggestions are addressed.

File: pennylane/devices/default_clifford.py (hunk 12 above)
Comment (obliviateandsurrender) [reply pennylane/devices/default_clifford.py:345]: Ahh, wait. I had the URL for the PennyLane measurements page here.
